### ISSUE TITLE
Org-mode blocks don't find R LSP server

### ISSUE THREAD

**Issue by wuqui** [emacs-lsp/lsp-mode#2004]
Hi, and thanks for this great package! I'm particularly happy that you have introduced `lsp-org` for integration into `org-mode`.

However, `lsp-org` doesn't work with my R code blocks when I call it inside a block. It says `No LSP server for R-mode(check *lsp-log*)`. Yet, there is no `lsp-log`, hence no backtrace here, sorry. R works fine with dedicated R buffers for me, and `lsp-org`works with Python org-mode blocks for me. Any idea?

Also, would this work with source blocks from different languages in one org-buffer? And can it intelligently make suggestions across source blocks sharing the same language? It seems that it can't help me with completion for packages that are imported in separate code blocks, for example. I know the two questions in this last paragraph are basically separate from the one above, but I didn't want to open three issues at the same time. If these are relevant enough, I could of course, though

**Comment by yyoncho** [comment]
> It says No LSP server for R-mode(check *lsp-log*). Yet, there is no lsp-log,

AFAICS r is register like that:
```
(lsp-register-client
 (make-lsp-client :new-connection (lsp-stdio-connection lsp-clients-r-server-command)
                  :major-modes '(ess-r-mode)
                  :server-id 'lsp-r))
```

At the same time org-mode thinks that the major more for that section is `R-mode`, you may add it to that block and eval it, and it should work(at least it will try to start). Alternatively, you may change the major mode for that block (but I don't know how).


> Also, would this work with source blocks from different languages in one
> org-buffer?

Yes.


> And can it intelligently make suggestions across source blocks
> sharing the same language?

Yes.

(or it is considered as bug)

**Comment by wuqui** [comment]
> > It says No LSP server for R-mode(check _lsp-log_). Yet, there is no lsp-log,
> 
> AFAICS r is register like that:
> 
> ```
> (lsp-register-client
>  (make-lsp-client :new-connection (lsp-stdio-connection lsp-clients-r-server-command)
>                   :major-modes '(ess-r-mode)
>                   :server-id 'lsp-r))
> ```
> 
> At the same time org-mode thinks that the major more for that section is `R-mode`, you may add it to that block and eval it, and it should work(at least it will try to start). Alternatively, you may change the major mode for that block (but I don't know how).

Ah, I found out that I need to start the code blocks with `#+BEGIN_SRC ess-r
` instead of the usual `R`. Now it works, nice!


> > Also, would this work with source blocks from different languages in one
> > org-buffer?
> 
> Yes.

That's really cool. With these features `org-mode` is just getting so much better than Jupyter notebooks etc :)

> > And can it intelligently make suggestions across source blocks
> > sharing the same language?

That's not working for me somehow. If I do the following, for example, I only get completions etc. in the first of the two blocks.

```
#+BEGIN_SRC jupyter-python 
import pandas as pd
lst = [1, 2]
#+END_SRC

#+BEGIN_SRC jupyter-python 
lst.append
pd.DataFr
#+END_SRC
```

Do I need to call `lsp-org` inside every block? Tried several combinations, nothing works. It seems to me that the most elegant soluation would be if `lsp-org` would just enable servers for all blocks (and their corresponding languages) across the whole buffer -- like a normal mode that applies to the whole buffer.

**Comment by yyoncho** [comment]
AFACS your org document lacks file information. Please use `test/fixtures/org-mode/demo.org` as a sample file for the required fields in org-mode headers.


> Do I need to call `lsp-org` inside every block? Tried several combinations,
> nothing works. It seems to me that the most elegant soluation would be if
> `lsp-org` would just enable servers for all blocks (and their corresponding
> languages) across the whole buffer -- like a normal mode that applies to the
> whole buffer.

ATM you have to call `lsp-org` in every block explicitly.

**Comment by yyoncho** [comment]
To elaborate a bit more: download demo.org. Then M-x lsp-org in first python block and when asked for project root select the current directory. Then do the same for the other python block. After doing that, the completion, navigation from the first file to the second one should work.

**Comment by wuqui** [comment]
Okay, thanks for your hints. I tried it, but doesn't work in the demo file for me either, unfortunately.

By the way, it might be a good idea to convert the example to Python 3 (`print X` > `print(x)`).

**Comment by yyoncho** [comment]
> Okay, thanks for your hints. I tried it, but doesn't work in the demo file for me either, unfortunately.

It is hard to diagnose the issue with that much info. Anyway, I will provide `emacs -q` demo with steps.

**Comment by wuqui** [comment]
> > Okay, thanks for your hints. I tried it, but doesn't work in the demo file for me either, unfortunately.
> 
> It is hard to diagnose the issue with that much info. Anyway, I will provide `emacs -q` demo with steps.

Yeah, sorry, but I don't really know what to do. Thanks a lot for your efforts!

I think fully working `lsp-org` could be really cool, given that `eglot` doesn't seem to have options for handling `org-babel` files and non-LSP options find it hard to deal with the above issues too.

**Comment by wuqui** [comment]
I have found that completion between blocks now works  as long as I'm inside the `org-edit-special` buffer of the current block -- completion will know about the types of variables that were defined in previous blocks like in my example above. It still doesn't work in the main org buffer, but that's already quite good 👍 

One thing, however, that does work in the 'normal' org buffer and _not_ in the `org-edit-special` buffer is `lsp-signature-activate`. It seems that all the ui and popup stuff doesnt' work in these buffers.

**Comment by yyoncho** [comment]
I don't use ess-r, jupiter-python or any of the mentioned technologies. Note that lsp-mode supports 50+ languages. So, if you want to get an answer, please provide a sample file, describe what you are doing, what you are expecting, and so on. And please, do that report as detailed as possible. Without this, it is really hard to help.

**Comment by gwbrck** [comment]
> > At the same time org-mode thinks that the major more for that section is `R-mode`, you may add it to that block and eval it, and it should work(at least it will try to start). Alternatively, you may change the major mode for that block (but I don't know how).
> 
> Ah, I found out that I need to start the code blocks with `#+BEGIN_SRC ess-r ` instead of the usual `R`. Now it works, nice!

Org-babel-execute no longer works for me then.

**Comment by KaiAragaki** [comment]
I have the same issue as @gwbrck - I can use a lsp with `ess-r` but can only execute chunks with `R`.
